src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1131,6 +1131,27 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
     }
     auto term = term_opt.value();
 
+    auto cfg = _metadata_cache.local().get_topic_cfg(model::tx_manager_nt);
+    if (!cfg) {
+        co_return init_tm_tx_reply{tx_errc::partition_not_exists};
+    }
+    if (stm->hosted_transactions_inited()) {
+        if (!stm->check_tx_membership(tx_id)) {
+            co_return tx_errc::not_coordinator;
+        }
+    } else {
+        auto hash_ranges_inited = co_await stm->try_init_hosted_transactions(
+          term, cfg->partition_count);
+        if (hash_ranges_inited != tm_stm::op_status::success) {
+            vlog(
+              txlog.info,
+              "got {} on initing hash ranges for tm {}",
+              hash_ranges_inited,
+              stm->get_partition());
+            co_return tx_errc::not_coordinator;
+        }
```

Comment:
The more I think about this the more I convinced this should go inside the `hosts` method discussed [there](https://github.com/redpanda-data/redpanda/pull/10181/files#r1183232112).

Changing the hosted txes is the stm wide action and it's very hard to prove that all possible interleaving with the existing tx_gateway_frontend's inflight operations don't cause trouble.

The only way to enforce it is to use rw-lock. If you check all tx_gateway_frontend's api are wrapped with `stm->read_lock()` so you need to wrap `try_init_hosted_transactions` with `write_lock`. But `do_init_tm_tx` already holds the read lock (deadlock!).

`hosts` is small enough method to juggling with the locks. You need to take read lock, sync and check if hosted txs are init-ed. If they do - use the ranges to check the tx_id. If they don't - release the read lock, take the write lock and init the hosted txes.